mobile/packages/swap/adapters/api/muesliswap/transformers.ts

```diff
@@ -132,22 +159,70 @@ export const transformersMaker = ({
         amountIn,
         amountOut,
         slippage,
-      }: Swap.EstimateRequest): QuoteRequest => ({
-        numbers_have_decimals: true,
-        sell_token: tokenIn,
-        buy_token: tokenOut,
-        ...(amountOut !== undefined && {buy_amount: String(amountOut)}),
-        ...(amountIn !== undefined && {sell_amount: String(amountIn)}),
-        ...(partner !== undefined && {partner}),
-        // muesli expects slippage as a percentage
-        slippage: slippage / 100,
-        excluded_sources: protocol
-          ? Object.values(Dex).filter(
-              (dex) =>
-                dex !== Dex.Unsupported && dex !== fromSwapProtocol(protocol),
+      }: Swap.EstimateRequest & {
+        routeHint?: {frontendOptions?: string[]}
+      }): QuoteRequest => {
+        // compute excluded_sources from Frontend Options when protocol is pinned and providers are available
+        let excluded_sources: ReadonlyArray<Dex> | Dex | undefined
+        if (
+          protocol !== undefined &&
+          providersCache?.dex_info &&
+          providersCache?.liquidity_source_info
+        ) {
+          const orderContract = fromSwapProtocol(protocol)
+          // find dex_info entry by matching key or scanning values
+          const dexEntry = providersCache.dex_info[orderContract]
+          if (dexEntry && Array.isArray(dexEntry.liquidity_protocols)) {
+            const allowedFO = dexEntry.liquidity_protocols
+              .map(
+                (lp: string) =>
+                  providersCache.liquidity_source_info?.[lp]?.frontend_option,
+              )
+              .filter((v: any) => typeof v === 'string')
+            const allFO = Object.values(
+              providersCache.liquidity_source_info ?? {},
             )
-          : (blockedProtocols?.map(fromSwapProtocol) ?? []),
-      }),
+              .map((v: any) => v?.frontend_option)
+              .filter((v: any) => typeof v === 'string')
+            const excludedFO = allFO.filter(
+              (fo: string) => !allowedFO.includes(fo),
+            )
+            excluded_sources = excludedFO as any
+          }
+        }
+
+        if (excluded_sources === undefined) {
+          if (providersCache?.liquidity_source_info && blockedProtocols) {
+            const mappedFO = blockedProtocols
+              .map(fromSwapProtocol)
+              .map(
+                (dex) =>
+                  (providersCache.liquidity_source_info as any)?.[dex]
+                    ?.frontend_option,
+              )
+              .filter((v): v is string => typeof v === 'string')
+            excluded_sources =
+              mappedFO.length > 0 ? (mappedFO as any) : undefined
```

Comment:
i think these any casts are AI thing (maybe the rules should be updated to avoid casting to any when there are types)